File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/fwmengine/world/ecs/components/ColliderComponent.java
package de.limbusdev.guardianmonsters.fwmengine.world.ecs.components;

import com.badlogic.ashley.core.Component;

import de.limbusdev.guardianmonsters.geometry.IntRect;


/**
 * Simple {@link Component} to hold an {@link IntRect} to represent a moving {@link com
 * .badlogic.ashley.core.Entity}'s collider.
 * Created by georg on 30.11.15.
 */
public class ColliderComponent implements Component {
    /* ............................................................................ ATTRIBUTES .. */
    public IntRect collider;
    /* ........................................................................... CONSTRUCTOR .. */
    public ColliderComponent(int x, int y, int width, int height) {
        this.collider = new IntRect(x,y,width,height);

    }
    /* ............................................................................... METHODS .. */
    
    /* ..................................................................... GETTERS & SETTERS .. */
}

File: GuardianMonsters/core/src/de/limbusdev/guardianmonsters/fwmengine/world/ecs/components/ColliderComponent.kt
package de.limbusdev.guardianmonsters.fwmengine.world.ecs.components

import com.badlogic.ashley.core.Component
import com.badlogic.ashley.core.Entity
import de.limbusdev.guardianmonsters.fwmengine.world.ecs.LimbusBehaviour

import de.limbusdev.utils.geometry.IntRect
import de.limbusdev.utils.geometry.IntVec2


/**
 * Simple [Component] to hold an [IntRect] to represent a moving [Entity]'s collider.
 *
 * @author Georg Eckert 2015-11-15
 */
class ColliderComponent
(
        enabled: Boolean = true,
        var offsetX: Int = 0,
        var offsetY: Int = 0,
        var width: Int = 16,
        var height: Int = 16,
        var inheritsExtent: Boolean = true,
        var isTrigger: Boolean = false
)
    : LimbusBehaviour()
{
    companion object
    {
        const val className ="ColliderComponent"
        val defaultJson = """
                    enabled: true,
                    offsetX: 0,
                    offsetY: 0,
                    width: 16,
                    height: 16,
                    inheritsExtent: true,
                    isTrigger: false
            """.trimMargin()
    }

    init
    {
        this.enabled = enabled
    }

    val asRectangle get() = IntRect((transform.x) + offsetX, (transform.y) + offsetY, width, height)

    /** Returns whether the given position is blocked by this collider */
    fun blocks(position : IntVec2) : Boolean
    {
        if(isTrigger) { return false }

        return asRectangle.contains(position)
    }
}
